tests/test_asgi.py

```diff
@@ -0,0 +1,262 @@
+
+
+class Sender:
+    def __init__(self):
+        self.events = []
+
+    async def __call__(self, event):
+        self.events.append(event)
+
+
+@pytest.fixture()
+def receive():
+    return Receiver()
+
+
+@pytest.fixture()
+def send():
+    return Sender()
+
+
+def test_asgiwhitenoise(loop, receive, send, method, whitenoise_application, files):
+    asgi_whitenoise = AsgiWhiteNoise(whitenoise_application, None)
+    scope = {
+        "type": "http",
+        "path": "/" + files.js_path,
+        "headers": [],
+        "method": method,
+    }
+    loop.run_until_complete(asgi_whitenoise(scope, receive, send))
+    assert receive.events == []
+    assert send.events[0]["status"] == 200
+    if method == "GET":
+        assert send.events[1]["body"] == files.js_content
+
+
+def test_asgiwhitenoise_not_found(
+    loop, receive, send, whitenoise_application, file_not_found
+):
+    asgi_whitenoise = AsgiWhiteNoise(whitenoise_application, file_not_found)
+    scope = {
+        "type": "http",
+        "path": "/static/foo.js",
+        "headers": [],
+        "method": "GET",
+    }
+    loop.run_until_complete(asgi_whitenoise(scope, receive, send))
+    assert receive.events == []
+    assert send.events == [
+        {"type": "http.response.start", "status": 404},
+        {"type": "http.response.body", "body": b"Not found"},
+    ]
+
+
+def test_asgiwhitenoise_not_http(
+    loop, receive, send, whitenoise_application, websocket
+):
+    asgi_whitenoise = AsgiWhiteNoise(whitenoise_application, websocket)
+    receive.events = [{"type": "websocket.connect"}]
+    scope = {
+        "type": "websocket",
+        "path": "/endpoint",
+        "headers": [],
+        "method": "GET",
+    }
+    loop.run_until_complete(asgi_whitenoise(scope, receive, send))
+    assert receive.events == []
+    assert send.events == [
+        {"type": "websocket.accept"},
+        {"type": "websocket.close"},
+    ]
+
+
+def test_serve_static_file(loop, send, method, block_size, static_file_sample):
+    loop.run_until_complete(
+        serve_static_file(
+            send, static_file_sample["static_file"], method, {}, block_size
+        )
+    )
+    expected_events = [
+        {
+            "type": "http.response.start",
+            "status": 200,
+            "headers": [
+                (b"last-modified", static_file_sample["modification_time"].encode()),
+                (b"etag", static_file_sample["static_file"].etag.encode()),
+                (b"content-length", str(static_file_sample["content_length"]).encode()),
+            ],
+        }
+    ]
+    if method == "GET":
+        for start in range(0, static_file_sample["content_length"], block_size):
+            expected_events.append(
+                {
+                    "type": "http.response.body",
+                    "body": static_file_sample["content"][start : start + block_size],
+                    "more_body": True,
+                }
+            )
```

Comment:
avoid using loops and complex constructs to compute test assertions  - instead just write out the expected values long form and assert against then